gcs/src/main/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopFileSystemBase.java

```diff
@@ -1483,6 +1488,29 @@ private Credential getCredential(Configuration config)
       }
     }
 
+    // If impersonation service account exists, then use current credential to request access token
+    // for the impersonating service account.
+      GoogleCloudStorageOptions options =
+          GoogleHadoopFileSystemConfiguration.getGcsFsOptionsBuilder(config)
+              .build()
+              .getCloudStorageOptions();
+    HttpRequestInitializer initializer =
+        new RetryHttpInitializer(
+            checkNotNull(credential, "credential must not be null"),
+            options.toRetryHttpInitializerOptions());
+      HttpTransport httpTransport =
+          HttpTransportFactory.createHttpTransport(
+              options.getTransportType(),
+              options.getProxyAddress(),
+              options.getProxyUsername(),
+              options.getProxyPassword());
```

Comment:
Is it possible to move these logic in the `credentialFromImpersonationServiceAccount` method?

Reply:
Just saw your earlier comment, what circular dependency do you mean?

Reply:
Yes, actually after a second thought just to get AT for SA no need to read all configs from this configuration at all. Some code in other credential factory in this package can be reused. The original intention to keep this here is to avoid circular import.